File: python/minilisp/parser.py
import re
from typing import List, Tuple

from minilisp import data

_SKIP_RE = re.compile(r'^(\s+|;.*)+')
_TOKEN_RE = re.compile(r'^[^\s);]+')
_NUM_RE = re.compile(r'^-?[0-9]+$')


class ParseError(Exception):
    pass


def _make_quote(value: data.Value) -> data.PairValue:
    return data.from_native_list([data.SymbolValue('quote'), value])


def _parse_skip(code: str) -> str:
    return _SKIP_RE.sub('', code)


def _parse_value(code: str) -> Tuple[data.Value, str]:
    if code.startswith('\''):
        value, rest_code = _parse_value(code[1:])
        return _make_quote(value), rest_code

    if code.startswith('('):
        values, rest_code = _parse_list(code[1:])
        rest_code = _parse_skip(rest_code)
        if not rest_code.startswith(')'):
            raise ParseError(rest_code)
        return data.from_native_list(values), rest_code[1:]

    match = _TOKEN_RE.search(code)
    if not match:
        raise ParseError(code)
    token = match.group()
    if _NUM_RE.search(token):
        value = data.IntegerValue(int(token))
    elif token == '#f':
        value = data.BooleanValue(False)
    elif token == '#t':
        value = data.BooleanValue(True)
    else:
        value = data.SymbolValue(token)
    return value, code[match.end():]


def _parse_list(code: str) -> Tuple[List[data.Value], str]:
    values = []
    while True:
        code = _parse_skip(code)
        if not code or code.startswith(')'):
            break
        # TODO: Support dots.
        value, code = _parse_value(code)
        values.append(value)
    return values, code


def parse(code: str) -> List[data.Value]:
    values, excess_code = _parse_list(code)
    if excess_code:
        raise ParseError(excess_code)
    return values

File: haskell/app/MiniLisp/Parser.hs
module MiniLisp.Parser( parse ) where

import Data.List
import Data.Maybe
import MiniLisp.Data
import Text.Regex

parseSkip :: String -> String
parseSkip code =
  let skipped = head $ fromJust $ matchRegex skipRe code in
  drop (length skipped) code
  where
    skipRe = mkRegex "(^([ \t\r\n]|;.*)*)"

makeQuote :: Value -> Value
makeQuote value =
  valueFromList [Symbol "quote", value]

parseValue :: String -> (Value, String)
parseValue code =
  case code of
    _ | "'" `isPrefixOf` code ->
      let (value, code') = parseValue (tail code) in
      (makeQuote value, code')
    _ | "(" `isPrefixOf` code ->
      let (values, code') = parseList (tail code) in
      let code'' = parseSkip code' in
      if ")" `isPrefixOf` code''
        then (valueFromList values, tail code'')
        else error $ "parse error: malformed list: " ++ code''
    _ ->
      case matchRegex tokenRe code of
        Nothing -> error $ "parse error: malformed token: " ++ code
        Just (token:_) ->
          let
            code' = drop (length token) code
            value = case token of
              _ | isJust (matchRegex numRe token) -> Integer (read token)
              "#t" -> Boolean True
              "#f" -> Boolean False
              _ -> Symbol token
          in
            (value, code')
        _ -> undefined
  where
    tokenRe = mkRegex "(^[^ \t\r\n);]+)"
    numRe = mkRegex "(-?[0-9]+$)"

parseList :: String -> ([Value], String)
parseList code =
  let code' = parseSkip code in
  case code' of
    "" -> ([], code')
    _ | ")" `isPrefixOf` code' -> ([], code')
    _ ->
      let (value, code'') = parseValue code' in
      let (values, code''') = parseList code'' in
      (value : values, code''')

parse :: String -> [Value]
parse code =
  let (exprs, code') = parseList code in
  if null code' then exprs else error $ "parse error: excessive code: " ++ code'
